ipq40xx: OpenMesh A42 (and similar devices) cleanups

ecsv opened the pull request:
Some minor cleanups of things which were noticed during the WPJ428 development. They don't change the behavior but are just here to keep up with the cleanups for other devices.

* DTS
  - use detailed reserved-memory regions for A42 (based on 864-07-dts-ipq4019-ap-dk01.1-c1-add-spi-and-ram-nodes.patch and input from Christian Lamparter/Roman Yeryomin)
  - use detailed reserved-memory regions for WPJ428 (based on 864-07-dts-ipq4019-ap-dk01.1-c1-add-spi-and-ram-nodes.patch and input from Christian Lamparter/Roman Yeryomin)
  - changed status values from "ok" to "okay" (all ipq40xx devices)
  - use constant to set GPIO_ACTIVE_HIGH for gpios (all ipq40xx devices)
  - remove unsupported property phy-reset-gpio for MR33
* image code
  - removed SUPPORTED_DEVICES for A42
  - added preferred DEVICE_DTS_CONFIG to A42
  - fixed AVM Fritz!Box 4040 dts filename to use the correct SoC name

ecsv committed:
ipq40xx: Remove SUPPORTED_DEVICES for A42

The SUPPORTED_DEVICES will be defined automatically via Device/Default in
ipq40xx.

Signed-off-by: Sven Eckelmann <sven.eckelmann@openmesh.com>

ecsv committed:
ipq40xx: Fix DTS status parameter values

The "Devicetree Speciﬁcation, Release v0.2 - 2.3.4 status" [1] only allows
the "okay" value for an operational device. The "ok" value will be accepted
by the kernel but should be avoided.

Signed-off-by: Sven Eckelmann <sven.eckelmann@openmesh.com>

ecsv committed:
ipq40xx: Provide prefered DTS config name for A42

The OpenMesh A42 will use the default config entry in the FIT when no other
on is found but prefers the config@om.a42. This only becomes relevant when
a Multi-FIT image is prepared for this device.

Signed-off-by: Sven Eckelmann <sven.eckelmann@openmesh.com>

ecsv committed:
ipq40xx: Use constant to set gpio active low/high

The GPIO configuration in the DTS have as third parameter the active
low/high configuration. This parameter is not easy to parse by humans when
it is only set to 0/1. It is better to use the predefined constants
GPIO_ACTIVE_HIGH and GPIO_ACTIVE_LOW.

Signed-off-by: Sven Eckelmann <sven.eckelmann@openmesh.com>

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-a42.dts:
```diff
@@ -169,7 +169,7 @@
 	pinctrl-names = "default";
 	status = "okay";
-	cs-gpios = <&tlmm 54 0>;
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>;
```
https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-a42.dts#L180

The pin_cs node sets
output-high;

If this is used to invert the gpio's polarity, wouldn't it be better to delete the pin_cs output-high property and just set the cs-gpios to GPIO_ACTIVE_LOW?

ecsv replied:
I already wanted to answer to this yesterday because I didn't get your point. But I've decided to sleep on it and hope that I get it today. Unfortunately, this didn't help :)

Let me try to explain it from my perspective:

* `output-high` in pin_cs is there to switch the gpio from in to output direction
   - high is only used to give it a defined initial state
   - low would also have "worked" in some sense because the spi driver changes it to whatever it wants anyway
* cs-gpios[0] with GPIO_ACTIVE_LOW should in theory not work correctly by itself because the slave would not have been selected correctly
  - spi-cs-high property would have been required to work around the wrong GPIO_ACTIVE_LOW (but why use a workaround when you can just set it correctly to GPIO_ACTIVE_HIGH)
  - in reality, something in the spi subsystem (or something else here) is just ignoring the parameter for polarity and only the spi-cs-high is applied in spi_set_cs

Let me just show it a couple of examples.

1.  current state:

        root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
         gpio54  : out 0 2mA no pull
        root@OpenWrt:/#  dmesg|grep -e spi0.0 |grep -v 'Kernel'
        [    0.551122] m25p80 spi0.0: w25q256 (32768 Kbytes)
        [    0.551187] 10 cmdlinepart partitions found on MTD device spi0.0
        [    0.554832] Creating 10 MTD partitions on "spi0.0":

2. output-high replaced with output-low:

        root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
         gpio54  : out 0 2mA no pull
        root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
        [    0.550952] m25p80 spi0.0: w25q256 (32768 Kbytes)
        [    0.551018] 10 cmdlinepart partitions found on MTD device spi0.0
        [    0.554661] Creating 10 MTD partitions on "spi0.0"


3.  with cs-gpios[0] set to GPIO_ACTIVE_LOW and output-high removed
    
        root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
         gpio54  : in  0 2mA no pull
        root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
        [    0.541523] m25p80 spi0.0: unrecognized JEDEC id bytes: 00, 00, 00
        [    0.548069] m25p80: probe of spi0.0 failed with error -2

4. spi-cs-high set for m25p80@0

        root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
         gpio54  : out 0 2mA no pull
        root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
        [    0.541811] m25p80 spi0.0: unrecognized JEDEC id bytes: 00, 00, 00
        [    0.548391] m25p80: probe of spi0.0 failed with error -2

Maybe it is just a misunderstanding from my side.

chunkeey replied:
> spi-cs-high property would have been required to work around the wrong GPIO_ACTIVE_LOW (but why use a workaround when you can just set it correctly to GPIO_ACTIVE_HIGH)
> in reality, something in the spi subsystem (or something else here) is just ignoring the parameter for polarity and only the spi-cs-high is applied in spi_set_cs

Looks like you are right about your assumption there. 
of_spi_register_master() only allocates a single int for the gpio pin/number:
https://github.com/torvalds/linux/blob/master/drivers/spi/spi.c#L2038
it only extracts the gpio number/pin from the property:
https://github.com/torvalds/linux/blob/master/drivers/spi/spi.c#L2049

And likewise spi_set_cs() only takes the spi-cs-high into account.

    static void spi_set_cs(struct spi_device *spi, bool enable)
    {
	if (spi->mode & SPI_CS_HIGH)
		enable = !enable;

	if (gpio_is_valid(spi->cs_gpio)) {
		gpio_set_value(spi->cs_gpio, !enable);

so yeah, the gpio flags in the cs-gpios property are ignored.
https://github.com/torvalds/linux/blob/master/drivers/spi/spi.c#L726

>   2. output-high replaced with output-low:
>    
>     root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
>      gpio54  : out 0 2mA no pull
>     root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
>     [    0.550952] m25p80 spi0.0: w25q256 (32768 Kbytes)
>     [    0.551018] 10 cmdlinepart partitions found on MTD device spi0.0
>     [    0.554661] Creating 10 MTD partitions on "spi0.0"
heh. ok so it's all about setting the gpio/pin to output then.

> 3. with cs-gpios[0] set to GPIO_ACTIVE_LOW and output-high removed
>
>    root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
>      gpio54  : in  0 2mA no pull
>    root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
>    [    0.541523] m25p80 spi0.0: unrecognized JEDEC id bytes: 00, 00, 00
>    [    0.548069] m25p80: probe of spi0.0 failed with error -2
>
Congratulations. I think you found a bona fide meta bug in the spi code there. 
it never set the gpio to output... which is definitely should since the
spi_set_cs() function does use gpio_set_value(). And gpio_set_value
assumes that the gpio is already set to output as per <https://lwn.net/Articles/532714/>.

the only saving grace I could find was this comment from 2008 in the vicinity:
https://github.com/torvalds/linux/blob/master/drivers/spi/spi.c#L540

	if (ctlr->cs_gpios)
		spi->cs_gpio = ctlr->cs_gpios[spi->chip_select];

	/* Drivers may modify this initial i/o setup, but will
	 * normally rely on the device being setup.  Devices
	 * using SPI_CS_HIGH can't coexist well otherwise...
	 */
	status = spi_setup(spi);

But yeah, I think the spi code should be fixed.

That said, I don't know what's better in this case now? This is could be very ugly and
simply doing any changes to this code could break existing setups (that used the wrong
values/settings). @CyrillePitchen

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4029-mr33.dts:
```diff
@@ -60,7 +60,7 @@
 			pinctrl-0 = <&mdio_pins>;
 			pinctrl-names = "default";
-			phy-reset-gpio = <&tlmm 47 0>;
+			phy-reset-gpio = <&tlmm 47 GPIO_ACTIVE_HIGH>;
```
Ram's ipq-mdio patch that introduced phy-reset-gpio was never applied :(. So for the time being this could be removed.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-a42.dts:
```diff
@@ -25,27 +25,6 @@
 	model = "OpenMesh A42";
 	compatible = "openmesh,a42", "qcom,ipq4019";
 
-	reserved-memory {
```
Or maybe the u-boot adds it? Can you check if the /reserved-memory/ node gets added in /proc/device-tree/ even if the defintion is removed from the .dts?

I know that writing to 87e00000 - 88000000 can cause the system to crash.
Especially addresses starting from 87e80000 onwards. As this memory belongs to the trustzone hypervisor thats running in the background the whole time. 

(It's kinda sad that the all CPUs are booted in the SVC mode and not HVP. As we miss out on the virtualization extensions because of this)

ecsv replied:
I've checked this previously on WPJ428 and it wasn't added. It should also not be added on A42 (but I cannot verify it right now from here). The region was removed from the WPJ428 because I couldn't show mkresin the bootup problem anymore which I had previously without it.

If this is really for the TZ then wouldn't it be better to have this reserved memory stored in a central place? Right now some IPQ40xx DTS add it and some not - and many of them add different regions.

chunkeey replied:
> I've checked this previously on WPJ428 and it wasn't added. It should also not be added on A42 (but I cannot verify it right now from here). The region was removed from the WPJ428 because I couldn't show mkresin the bootup problem anymore which I had previously without it.

I remember the grief tracing down the booting issue on the WPJ428 with @mehlis . I don't know if this is related or not. But back the WPJ428 would simply not boot (and not leave any messages on the serial console) and no matter what we tried it remained silent. Only after booting a modified rt-ac58u uboot the generated linux initramfs image from the LEDE-IPQ40XX would run. The reason why this didn't work was because the Compex u-boot would place the processed fdt into the reserved-memory area (since the initramfs image was slightly too big). And this prevented the kernel from parsing the fdt (apart from the early_fdt). Once the reserved-memory node (rsvd1@87000000) was removed the device booted the images... Since @mehlis didn't want to remove the reserve-memory node he added the "set fdt_high 83000000" u-boot env variable, so this was no longer a problem.


About the reserved-memory node. I managed to get the information from @yeryomin on the mailing-list. He was kind enough to make a patch with the proper memory node names.
This was added to:

https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/patches-4.14/864-07-dts-ipq4019-ap-dk01.1-c1-add-spi-and-ram-nodes.patch

If you want you could run this experiment in the u-boot on the WPJ428 and see if you get the same result (I c&p this from my RT-AC58U with the stock u-boot) :
    (IPQ40xx) # mw 87e70000 0

    (IPQ40xx) # md 87e80000
    87e80000: e59ff364 e59ff364 e59ff364 e59ff364    d...d...d...d...
    87e80010: e59ff364 e59ff364 e59ff360 e59ff35c    d...d...`...\...
    87e80020: ee0d0f50 ee0def90 e59f0354 e12fff30    P.......T...0./.
    87e80030: e3500000 1a0000ba ee1d0f50 e59f3344    ..P.....P...D3..
    87e80040: e5933000 e3530000 0a000086 eaffffff    .0....S.........
    87e80050: e24f0008 e59f1330 e1500001 1afffffe    ..O.0.....P.....
    87e80060: e59f1328 e5911000 ee100fb0 e2000cff    (...............
    87e80070: e1a00440 e0000190 ee101fb0 e20110ff    @...............
    87e80080: e0810000 e59f1308 e3a02001 e7812100    ......... ...!..
    87e80090: e59f1300 ee102fb0 e20220ff e2822001    ...../... ... ..
    87e800a0: e1910f9f e3500000 01810f92 03500000    ......P.......P.
    87e800b0: 1afffffa f57ff05f ee100fb0 e2000cff    ...._...........
    87e800c0: e1a00440 e59f12d0 e7912100 e2422001    @........!... B.
    87e800d0: e7812100 e59f12b4 e5911000 ee100fb0    .!..............
    87e800e0: e2000cff e1a00440 e0000190 ee101fb0    ....@...........
    87e800f0: e20110ff e0810000 e59f12a4 e7911100    ................
    (IPQ40xx) # mw 87e80000 0
    data abort  
    pc : [<873149f8>]          lr : [<87308578>]
    sp : 86edfc28  ip : 86ef4412     fp : 00000000
    r10: 00000000  r9 : 00000000     r8 : 86edff68
    r7 : 00000003  r6 : 8737e624     r5 : 86ef4420  r4 : 8736c154
    r3 : 00000000  r2 : 00000010     r1 : 00000000  r0 : 00000000
    Flags: nZCv  IRQs off  FIQs off  Mode SVC_32
    Resetting CPU ...

    resetting ...

> If this is really for the TZ then wouldn't it be better to have this reserved memory stored in a central place? Right now some IPQ40xx DTS add it and some not - and many of them add different regions.

Some u-boots do read/add or fiddle with the reserved-memory / memory nodes.
https://github.com/riptidewave93/meraki-uboot/blob/mr33-20170427/board/qcom/ipq40xx_cdp/ipq40xx_cdp.c#L764
(depends on whenever CONFIG_QCA_APPSBL_DLOAD is set or not).

so the question becomes: can the reserved-memory nodes be stored in centrally?
or will this lead to devices needing /delete-node/ rsvd0; hacks?

(Also changing the qcom-ipq4019.dtsi is difficult since the file is upstream, so any patches to the file should be sent to linux-arm-msm@vger.kernel.org as well)

yeryomin replied:
Theoretically, different boards may have different reserved memory map set.
That's why I didn't add that full set into common ipq40xx dtsi.
Practically, only the time will show...

ecsv replied:
Both WPJ428 and A42 crashed with the mw and both didn't create the reserved memory entries in the device tree. Now comes the biggest question - how to find out what memory regions are reserved for what.

And to the u-boot code. The meraki code also doesn't create the node but just adjusts it. I have also checked the A42 and it would use the same code when CONFIG_QCA_APPSBL_DLOAD would have been enabled. It transforms (in theory):

           reserved-memory {
                   #address-cells = <0x1>;
                   #size-cells = <0x1>;
                   ranges;
    
                  /* apps_bl, sbl? */
                   rsvd1@87000000 {
                           reg = <0x87000000 0x500000>;
                           no-map;
                   };
    
                   /* cnss_debug? */
                   wifi_dump@87500000 {
                           reg = <0x87500000 0x600000>;
                           no-map;
                   };
    
                   /* cpu_context_dump, tz_apps, smem, tz? */
                   rsvd2@87B00000 {
                           reg = <0x87b00000 0x500000>;
                           no-map;
                   };
           };

To

           reserved-memory {
                   #address-cells = <0x1>;
                   #size-cells = <0x1>;
                   ranges;
    
                   /* cpu_context_dump, tz_apps, smem, tz? */
                   rsvd1@87b00000 {
                           reg = <0x87b00000 0x500000>;
                           no-map;
                   };
           };

I find it also quite interesting that all these nodes + reset_crashdump are necessary to reach reset_crashdump. But since none of the code is actually active, it doesn't seem to be important at all.

I will remove this commit now but this is definitely something where we are missing some information to make a decision about how to handle it.

ecsv replied:
@yeryomin could you please check https://github.com/openwrt/openwrt/pull/810/commits/7dcbd3180d6f74427b0a3d1ff3c446c6f390c1bd ("ipq40xx: Add reserved memory for WPJ428") makes sense? You seem to know a little bit more about it then me.

chunkeey replied:
> I find it also quite interesting that all these nodes + reset_crashdump are necessary to reach reset_crashdump. But since none of the code is actually active, it doesn't seem to be important at all.

>I will remove this commit now but this is definitely something where we are missing some information to make a decision about how to handle it.

Just a (related) note about tz: The tz can be accessed through the qcom-scm.c
https://github.com/torvalds/linux/blob/master/drivers/firmware/qcom_scm.c.
Nowadays, it would only require adding the scm firmware node in the dts.

I used to have this enabled:
https://github.com/chunkeey/LEDE-IPQ40XX/commit/4c75910103f2fbe934f299d0bade25226feab281#diff-75a9ee647eceaa46bea8e9697abdfee2

And this firmware node was also part of the https://patchwork.kernel.org/patch/10189259/
that became
https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/patches-4.14/101-ARM-dts-ipq4019-Add-a-few-peripheral-nodes.patch
However, the node has been removed from the patch by I think @mkresin (but it could also have been  john?). so there's probably no point in adding it back.

yeryomin replied:
@ecsv it makes sense for sure

chunkeey reviewed (commented):
just nitpicks.

That said, you reminded me to change the fritz4040 filename. As it is a ipq4018 as well.

ecsv committed:
ipq40xx: Adjust SoC name of AVM Fritz!Box 4040

The AVM Fritz!Box 4040 uses an IPQ4018 as SoC and not an IPQ4019. The DTS
must be adjusted to reflect this.

Signed-off-by: Sven Eckelmann <sven@narfation.org>

ecsv committed:
ipq40xx: Remove phy reset gpio from Cisco Meraki MR33

There is currently no code to read the phy reset gpios for the ethernet
PHY. It would also have been better to use the more common name
"phy-reset-gpios" for this property.

Signed-off-by: Sven Eckelmann <sven@narfation.org>

ecsv committed:
ipq40xx: Use detailed reserved memory for A42

The APPSBL and anything after that it not allowed to touch some of the
memory regions which are used by other components. Still trying to write to
the memory can lead to sudden device restarts

  (IPQ40xx) # mw 87e80000 0
  data abort
  pc : [<873149f8>]          lr : [<87308578>]
  sp : 86edfc28  ip : 86ef4412     fp : 00000000
  r10: 00000000  r9 : 00000000     r8 : 86edff68
  r7 : 00000003  r6 : 8737e624     r5 : 86ef4420  r4 : 8736c154
  r3 : 00000000  r2 : 00000010     r1 : 00000000  r0 : 00000000
  Flags: nZCv  IRQs off  FIQs off  Mode SVC_32
  Resetting CPU ...

The device manufacturer only provided a very rough list of regions:

* rsvd1:     0x87000000 0x500000
* wifi_dump: 0x87500000 0x600000
* rsvd2:     0x87b00000 0x500000

A more detailed list for devices using the AP.DK reference design memory
maps was provided by Roman Yeryomin <roman@advem.lv> in commit 56f2df879fd
("ipq806x: ipq4019: add ap-dk01.1-c1 board support"):

* apps_bl:          0x87000000 0x400000
* sbl:              0x87400000 0x100000
* cnss_debug:       0x87500000 0x600000
* cpu_context_dump: 0x87b00000 0x080000
* tz_apps:          0x87b80000 0x280000
* smem:             0x87e00000 0x080000
* tz:               0x87e80000 0x180000

The u-boot function ipq_fdt_mem_rsvd_fixup seems to suggest that only the
rsvd2 (tz_apps, smem, tz) should be protected. All other regions would have
been removed by it when CONFIG_QCA_APPSBL_DLOAD is not enabled. This allows
to reduce the 16MB reserved memory region to only 4.5MB.

Signed-off-by: Sven Eckelmann <sven.eckelmann@openmesh.com>

ecsv committed:
ipq40xx: Add reserved memory for WPJ428

The APPSBL and anything after that it not allowed to touch some of the
memory regions which are used by other components. Still trying to write to
the memory can lead to sudden device restarts

  (IPQ40xx) # mw 87e80000 0
  data abort
  pc : [<873149f8>]          lr : [<87308578>]
  sp : 86edfc28  ip : 86ef4412     fp : 00000000
  r10: 00000000  r9 : 00000000     r8 : 86edff68
  r7 : 00000003  r6 : 8737e624     r5 : 86ef4420  r4 : 8736c154
  r3 : 00000000  r2 : 00000010     r1 : 00000000  r0 : 00000000
  Flags: nZCv  IRQs off  FIQs off  Mode SVC_32
  Resetting CPU ...

The device manufacturer only provided a very rough list of regions:

* rsvd1:     0x87000000 0x500000
* wifi_dump: 0x87500000 0x600000
* rsvd2:     0x87b00000 0x500000

A more detailed list for devices using the AP.DK reference design memory
maps was provided by Roman Yeryomin <roman@advem.lv> in commit 56f2df879fd
("ipq806x: ipq4019: add ap-dk01.1-c1 board support"):

* apps_bl:          0x87000000 0x400000
* sbl:              0x87400000 0x100000
* cnss_debug:       0x87500000 0x600000
* cpu_context_dump: 0x87b00000 0x080000
* tz_apps:          0x87b80000 0x280000
* smem:             0x87e00000 0x080000
* tz:               0x87e80000 0x180000

The u-boot function ipq_fdt_mem_rsvd_fixup seems to suggest that only the
rsvd2 (tz_apps, smem, tz) should be protected. All other regions would have
been removed by it when CONFIG_QCA_APPSBL_DLOAD is not enabled. This allows
to reduce the 16MB reserved memory region to only 4.5MB.

Signed-off-by: Sven Eckelmann <sven.eckelmann@openmesh.com>

ecsv commented:
Have rebased everything because tango.drone.openwrt.ninja just didn't want to checkout the old state. Also pressing "restart" had no effect. But it looks like a rebase (and thus change of the commit ids) was enough to trigger a new check.

chunkeey reviewed (approved):
Thanks!

(And thanks again, for the FB4040 patch)

mkresin commented:
Excellent job from everyone. Consider it as merged.

> MKR33:
> Ram's ipq-mdio patch that introduced phy-reset-gpio was never applied :(
> However, the node has been removed from the patch by I think @mkresin (but it could also have been @blogic ?). so there's probably no point in adding it back.

Yeah, I'm the one who is to blame. I dropped everything from the split/kernel update commit that wasn't necessarly required, to not introduce new features or changes within the split/kernel update topic. It was never meant as reject of the patch. It was just the wrong time/topic for the patch.

> WPJ428
> Since @mehlis didn't want to remove the reserve-memory node he added the "set fdt_high 83000000" u-boot env variable, so this was no longer a problem.

`set fdt_high 83000000` is set by the bootipq command as well. This information is kind of offtopic but the fdt_high approach is correct

> Some u-boots do read/add or fiddle with the reserved-memory / memory nodes.

It is exactly what I noticed with the ipq806x WPQ864. The bootloader of that board takes care of adding a single 16MB (?) reserved node. That was the reason for the question if a reserved is required for the WPJ428.

I'm still not sure if we really need the wifi_dump/cnss_debug node. I would expect that such a node need to be passed to driver/firmware, similar to the smem node on ipq806x. But I won't bet against a hardcoded memory region somewhere. The far easier explaination is that I still miss something.

This PR/commits is the best explaination for the reserved memory regions I found so far. I never guessed that tz stands for trust zone.

mkresin commented:
Merged thanks!
